include/appbase/application_instance.hpp

```diff
@@ -0,0 +1,103 @@
+#pragma once
+
+namespace appbase {
+
+static application& app();
+
+// ------------------------------------------------------------------------------------------
+template <typename Impl>
+class plugin : public abstract_plugin {
+public:
+   plugin() : _name(boost::core::demangle(typeid(Impl).name())) {}
+   virtual ~plugin() {}
+
+   virtual state get_state() const final {
+      return _state;
+   }
+   virtual const std::string& name() const final {
+      return _name;
+   }
+
+   virtual void register_dependencies() {
+      static_cast<Impl*>(this)->plugin_requires([&](auto& plug) {});
+   }
+
+   virtual void initialize(const variables_map& options) final {
+      if (_state == registered) {
+         _state = initialized;
+         static_cast<Impl*>(this)->plugin_requires([&](auto& plug) { plug.initialize(options); });
+         static_cast<Impl*>(this)->plugin_initialize(options);
+         // ilog( "initializing plugin ${name}", ("name",name()) );
+         app().plugin_initialized(*this);
+      }
+      assert(_state == initialized); /// if initial state was not registered, final state cannot be initialized
+   }
+
+   virtual void handle_sighup() override {}
+
+   virtual void startup() final {
+      if (_state == initialized) {
+         _state = started;
+         static_cast<Impl*>(this)->plugin_requires([&](auto& plug) { plug.startup(); });
+         static_cast<Impl*>(this)->plugin_startup();
+         app().plugin_started(*this);
+      }
+      assert(_state == started); // if initial state was not initialized, final state cannot be started
+   }
+
+   virtual void shutdown() final {
+      if (_state == started) {
+         _state = stopped;
+         // ilog( "shutting down plugin ${name}", ("name",name()) );
+         static_cast<Impl*>(this)->plugin_shutdown();
+      }
+   }
+
+protected:
+   plugin(const string& name) : _name(name) {}
+
+private:
+   state _state = abstract_plugin::registered;
+   std::string _name;
+};
+
+// ------------------------------------------------------------------------------------------
+template <typename Data, typename DispatchPolicy>
+void channel<Data, DispatchPolicy>::publish(int priority, const Data& data) {
+   if (has_subscribers()) {
+      // this will copy data into the lambda
+      app().executor().post(priority, [this, data]() { _signal(data); });
```

Comment:
You are depending on the executor having a post that takes a priority and a lambda. So I think you can just require that and not need the `post_cb` std::function.

Reply:
Same answer as above comment.